src/RateLimiter/Counter.php

```diff
@@ -0,0 +1,102 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Yii\Web\RateLimiter;
+
+use Psr\Http\Message\ServerRequestInterface;
+
+final class Counter
+{
+    private int $interval = 360;
+
+    private ?string $id = null;
+
+    /**
+     * @var callable
+     */
+    private $idCallback;
+
+    private bool $init = false;
+
+    private CounterStorageInterface $storage;
+
+    public function __construct(CounterStorageInterface $storage)
+    {
+        $this->storage = $storage;
+    }
+
+    public function init(ServerRequestInterface $request): self
+    {
+        $this->id = $this->generateId($request);
+
+        if (!$this->storage->has($this->id)) {
+            $this->storage->set($this->id, 0, $this->interval);
+        }
+
+        $this->init = true;
+
+        return $this;
+    }
+
+    public function setIdCallback(callable $callback): self
+    {
+        $this->idCallback = $callback;
+
+        return $this;
+    }
+
+    public function setId(string $id): self
+    {
+        $this->id = $id;
+
+        return $this;
+    }
+
+    public function setInterval(int $interval): self
+    {
+        $this->interval = $interval;
+
+        return $this;
+    }
+
+    public function increment(): void
+    {
+        $value = $this->getCounterValue();
+        $value++;
+
+        $this->storage->set($this->id, $value, $this->interval);
```

Comment:
While it makes code simple, this rate limiter implementation (fixed window) results in possibility of traffic spikes on the edge of the interval. The behavior for interval = 360 (6 minutes) and limit = 10000 is that after 10000 requests there will be 504 HTTP response for exactly 6 minutes. Then there could be quick 10000 requests again and another 6 minute pause etc. So you'll still have traffic spikes this way. That's a no go for a good rate limiter.

Good algorithms preventing traffic spikes are:

1. Leaky bucket https://en.wikipedia.org/wiki/Leaky_bucket. That one is in Yii 2.
2. Sliding window.

See:

1. https://blog.cloudflare.com/counting-things-a-lot-of-different-things/
2. https://habr.com/ru/post/448438/

Reply:
I'd implement GCRA:

- https://brandur.org/rate-limiting
- https://blog.ian.stapletoncordas.co/2018/12/understanding-generic-cell-rate-limiting.html

Reply:
And counts are lost. for example
1. A request get counter value (X)
2. B request get counter value (X)
3. A requrest set counter value (X+1)
4. B request set counter value (X+1)

In fact, X + 2. In multi server environments, this can mean a lot of lost counting. 
I recommend that the storage interface have a `incrementAndGet()` method. [Redis](https://redis.io/commands/incr), mysql allow counting in atomic.

Reply:
Yes, increment operation should be atomic.